tutor/src/components/annotations/annotation.jsx

```diff
@@ -147,78 +140,118 @@ export default class AnnotationWidget extends React.Component {
+    };
+  }
+
   componentDidMount() {
+    if (this.props.windowImpl.location.hash) {
+      this.setupPendingHighlightScroll(this.props.windowImpl.location.hash);
+    }
+
     this.handleSelectionChange = debounce(() => {
       if (!this.activeHighlight) {
         this.setWidgetStyle();
       }
     }, 80);
-    this.props.windowImpl.document.addEventListener('selectionchange', this.handleSelectionChange);
+
+    when(
+      () => User.annotations.any,
+      () => {
+        this.props.windowImpl.document.addEventListener('selectionchange', this.handleSelectionChange);
+        this.initializePage();
+      },
+    );
   }
 
   componentWillReceiveProps(nextProps) {
-    this.needToGetReferenceElements = nextProps.documentId !== this.props.documentId;
-    if (this.needToGetReferenceElements) {
-      this.widgetStyle = null;
-      this.activeHighlight = null;
-    }
+    if (nextProps.documentId !== this.props.documentId) {
+      this.initializePage();
+    };
+    this.widgetStyle = null;
+    this.activeHighlight = null;
   }
 
-  componentDidUpdate() {
-    this.getReferenceElements();
-    if (this.savedSelection) {
-      const selection = this.props.windowImpl.getSelection();
+  // componentDidUpdate() {
+  //   this.getReferenceElements();
+  //   if (this.savedSelection) {
+  //     const selection = this.props.windowImpl.getSelection();
 
-      if (selection.isCollapsed) {
-        this.savedSelection.restore();
-      }
-    }
+  //     if (selection.isCollapsed) {
+  //       this.savedSelection.restore();
+  //     }
+  //   }
 
-    const navElements = document.querySelectorAll([
-      '.center-panel',
-      '.reading-content'
-    ].join(','));
+  //   const navElements = document.querySelectorAll([
+  //     '.center-panel',
+  //     '.reading-content'
+  //   ].join(','));
 
-    for (const el of navElements) {
-      el.style.display = this.showWindowShade ? 'none' : '';
-    }
+  //   for (const el of navElements) {
+  //     el.style.display = this.showWindowShade ? 'none' : '';
+  //   }
 
-  }
+  // }
 
   componentWillUnmount() {
     this.props.windowImpl.document.removeEventListener('selectionchange', this.handleSelectionChange)
   }
 
+  initializePage() {
+    this.getReferenceElements();
+    imagesComplete(
+      this.props.windowImpl.document.querySelector('.book-content')
+    ).then(() => {
+      this.annotationsForThisPage.forEach((annotation) => {
+        this.highlightEntry(annotation);
+      });
+      this.canRenderSidebarButtons = true;
+      if (this.scrollToPendingAnnotation) {
+        this.scrollToPendingAnnotation();
+      }
+    });
+  }
+
   @action.bound
   highlightEntry(entry) {
     const selection = this.restoreSelectionWithReferenceId(entry.selection);
-
     if (selection) {
       const rect = getSelectionRect(this.props.windowImpl, selection);
-
       entry.style = {
         top: rect.top - this.parentRect.top,
-        position: 'absolute'
+        position: 'absolute',
```

Comment:
Do you want the trailing comma?

Reply:
Yeah, our (my) preferred eslint config uses trailing comma on objects.  I like it because now if I add another line the git diff doesn't have to pollute the previous line by adding a comma.